File: backend/screener.py
import pandas as pd
import numpy as np


def apply_filters(df, per_max=15.0, pbr_max=1.5, roe_min=10.0, rsi_max=40.0, w52_min=-20.0):
    if df is None or df.empty:
        return pd.DataFrame()

    mask = pd.Series(True, index=df.index)

    if "PER" in df.columns:
        mask &= df["PER"].notna() & (df["PER"] > 0) & (df["PER"] <= per_max)
    if "PBR" in df.columns:
        mask &= df["PBR"].notna() & (df["PBR"] > 0) & (df["PBR"] <= pbr_max)
    if "ROE" in df.columns:
        mask &= df["ROE"].notna() & (df["ROE"] >= roe_min)
    if "RSI" in df.columns:
        mask &= df["RSI"].notna() & (df["RSI"] <= rsi_max)
    if "52W_change" in df.columns:
        mask &= df["52W_change"].notna() & (df["52W_change"] <= w52_min)

    return df[mask].copy()


def _normalize(series, higher_is_better=False):
    s_min, s_max = series.min(), series.max()
    if s_max == s_min:
        return pd.Series(50.0, index=series.index)
    norm = (series - s_min) / (s_max - s_min) * 100
    return norm if higher_is_better else (100 - norm)


def calculate_score(df):
    if df is None or df.empty:
        return df
    df = df.copy()

    weights = {
        "PER":        (0.25, False),
        "PBR":        (0.20, False),
        "ROE":        (0.20, True),
        "RSI":        (0.20, False),
        "52W_change": (0.15, False),
    }

    score = pd.Series(0.0, index=df.index)
    for col, (w, higher) in weights.items():
        if col in df.columns and df[col].notna().sum() > 1:
            score += _normalize(df[col], higher_is_better=higher) * w

    df["score"] = score.round(1)
    return df

File: frontend/src/screener.ts
import { Stock, Filters } from './types'

export function applyFilters(stocks: Stock[], f: Filters): Stock[] {
  return stocks.filter(s => {
    // 값이 null이면 해당 항목은 필터 통과 (데이터 없음 = 필터 적용 불가)
    if (s.PER !== null && (s.PER <= 0 || s.PER > f.per_max)) return false
    if (s.PBR !== null && (s.PBR <= 0 || s.PBR > f.pbr_max)) return false
    if (s.ROE !== null && s.ROE < f.roe_min) return false
    if (s.RSI !== null && s.RSI > f.rsi_max) return false
    if (s['52W_change'] !== null && s['52W_change'] > f.w52_min) return false
    // PER은 필수: 아예 없으면 제외
    if (s.PER === null) return false
    return true
  })
}

function normalize(values: number[], higherIsBetter: boolean): number[] {
  const min = Math.min(...values)
  const max = Math.max(...values)
  if (max === min) return values.map(() => 50)
  return values.map(v => {
    const n = ((v - min) / (max - min)) * 100
    return higherIsBetter ? n : 100 - n
  })
}

const WEIGHTS: { key: keyof Stock; weight: number; higherIsBetter: boolean }[] = [
  { key: 'PER',        weight: 0.25, higherIsBetter: false },
  { key: 'PBR',        weight: 0.20, higherIsBetter: false },
  { key: 'ROE',        weight: 0.20, higherIsBetter: true  },
  { key: 'RSI',        weight: 0.20, higherIsBetter: false },
  { key: '52W_change', weight: 0.15, higherIsBetter: false },
]

export function calculateScores(stocks: Stock[]): (Stock & { score: number })[] {
  if (stocks.length === 0) return []

  const scores = new Array(stocks.length).fill(0)

  for (const { key, weight, higherIsBetter } of WEIGHTS) {
    const vals = stocks.map(s => s[key] as number | null)
    const valids = vals.filter((v): v is number => v !== null)
    if (valids.length < 2) continue

    const normed = normalize(valids, higherIsBetter)
    let ni = 0
    for (let i = 0; i < stocks.length; i++) {
      if (vals[i] !== null) {
        scores[i] += normed[ni++] * weight
      }
    }
  }

  return stocks.map((s, i) => ({ ...s, score: Math.round(scores[i] * 10) / 10 }))
}
